panic on podman push

When executing `podman push registry.mydomain.com/imagename`, I get the following error:
```
Getting image source signatures
Copying blob 3a8defd3c960 [=======>------------------------------] 10.0MiB / 50.0MiB
Copying blob 002555c37f66 [=============>------------------------] 6.0MiB / 16.4MiB
Copying blob 831b66a484dc [=>------------------------------------] 7.3MiB / 113.7MiB
Copying blob 9cc9add37421 [>-------------------------------------] 7.3MiB / 475.4MiB
Copying blob ccf71bd6dffe skipped: already exists
Copying blob f256d8ec889c skipped: already exists
Copying blob e1ca698764a7 [--------------------------------------] 8b / 8.5MiB
Copying blob 6aad20fe678c [--------------------------------------] 8b / 2.2MiB
panic: d.nx != 0

goroutine 86 [running]:
crypto/sha256.(*digest).checkSum(0xc0007d8b40, 0x0, 0x0, 0x0, 0x0)
	/usr/lib/go-1.11/src/crypto/sha256/sha256.go:253 +0x1e9
crypto/sha256.(*digest).Sum(0xc0003fe980, 0x0, 0x0, 0x0, 0xc0007d8c68, 0xf98d1219, 0xb1cdc63ac2ffd3e2)
	/usr/lib/go-1.11/src/crypto/sha256/sha256.go:229 +0x69
github.com/opencontainers/go-digest.NewDigest(0x191b791, 0x6, 0x1b7c4c0, 0xc0003fe980, 0x18, 0xc0007d8c60)
	vendor/github.com/opencontainers/go-digest/digest.go:40 +0x46
github.com/opencontainers/go-digest.(*digester).Digest(0xc00056ca80, 0x1b6ccd8, 0xc0006feec0)
	vendor/github.com/opencontainers/go-digest/digester.go:38 +0x49
github.com/containers/image/v5/docker.(*dockerImageDestination).PutBlob(0xc000274de0, 0x1b7a040, 0xc00003e138, 0x1b62680, 0xc00036a068, 0x0, 0x0, 0xffffffffffffffff, 0x0, 0x0, ...)
	vendor/github.com/containers/image/v5/docker/docker_image_dest.go:169 +0x669
github.com/containers/image/v5/copy.(*copier).copyBlobFromStream(0xc0001048f0, 0x1b7a040, 0xc00003e138, 0x7fb12c201238, 0xc0003c89a0, 0xc0004be550, 0x47, 0x1068400, 0x0, 0x0, ...)
	vendor/github.com/containers/image/v5/copy/copy.go:1167 +0x40b
github.com/containers/image/v5/copy.(*imageCopier).copyLayerFromStream(0xc000275260, 0x1b7a040, 0xc00003e138, 0x7fb12c201238, 0xc0003c89a0, 0xc0004be550, 0x47, 0x1068400, 0x0, 0x0, ...)
	vendor/github.com/containers/image/v5/copy/copy.go:1030 +0x191
github.com/containers/image/v5/copy.(*imageCopier).copyLayer(0xc000275260, 0x1b7a040, 0xc00003e138, 0xc0004be550, 0x47, 0x1068400, 0x0, 0x0, 0x0, 0x0, ...)
	vendor/github.com/containers/image/v5/copy/copy.go:975 +0x39f
github.com/containers/image/v5/copy.(*imageCopier).copyLayers.func1(0x2, 0xc0004be550, 0x47, 0x1068400, 0x0, 0x0, 0x0, 0x0, 0xc000386300, 0x26, ...)
	vendor/github.com/containers/image/v5/copy/copy.go:755 +0x3ea
created by github.com/containers/image/v5/copy.(*imageCopier).copyLayers.func2
	vendor/github.com/containers/image/v5/copy/copy.go:766 +0x1b5
```
Even though the error occurs in podman, I opened the issue here because the backtrace suggests that the issue occurs in the `containers/image` code.

The registry is a normal docker registry.

## System Information
```
OS: Debian Buster
Podman Version: Occurs on both version 1.6.2 (installed via ubuntu PPA) and source build of podman (1.6.4-dev)
```

--- a/docker/docker_image_dest.go
+++ b/docker/docker_image_dest.go
@@ -170,7 +170,7 @@ func (d *dockerImageDestination) PutBlob(ctx context.Context, stream io.Reader, 
 		defer uploadReader.Terminate(errors.New("Reading data from an already terminated upload"))
 		res, err = d.c.makeRequestToResolvedURL(ctx, "PATCH", uploadLocation.String(), map[string][]string{"Content-Type": {"application/octet-stream"}}, uploadReader, inputInfo.Size, v2Auth, nil)
 		if err != nil {
-			logrus.Debugf("Error uploading layer chunked, response %#v", res)
+			logrus.Debugf("Error uploading layer chunked %v", err)
 			return nil, err
 		}
 		defer res.Body.Close()

--- a/docker/docker_image_dest.go
+++ b/docker/docker_image_dest.go
@@ -174,6 +174,9 @@ func (d *dockerImageDestination) PutBlob(ctx context.Context, stream io.Reader, 
 			return nil, err
 		}
 		defer res.Body.Close()
+		if !successStatus(res.StatusCode) {
+			return nil, errors.Wrapf(client.HandleErrorResponse(res), "Error uploading layer chunked")
+		}
 		uploadLocation, err := res.Location()
 		if err != nil {
 			return nil, errors.Wrap(err, "Error determining upload URL")

--- a/docker/docker_image_dest.go
+++ b/docker/docker_image_dest.go
@@ -16,6 +16,7 @@ import (
 
 	"github.com/containers/image/v5/docker/reference"
 	"github.com/containers/image/v5/internal/iolimits"
+	"github.com/containers/image/v5/internal/uploadreader"
 	"github.com/containers/image/v5/manifest"
 	"github.com/containers/image/v5/pkg/blobinfocache/none"
 	"github.com/containers/image/v5/types"
@@ -162,19 +163,27 @@ func (d *dockerImageDestination) PutBlob(ctx context.Context, stream io.Reader, 
 
 	digester := digest.Canonical.Digester()
 	sizeCounter := &sizeCounter{}
-	tee := io.TeeReader(stream, io.MultiWriter(digester.Hash(), sizeCounter))
-	res, err = d.c.makeRequestToResolvedURL(ctx, "PATCH", uploadLocation.String(), map[string][]string{"Content-Type": {"application/octet-stream"}}, tee, inputInfo.Size, v2Auth, nil)
+	uploadLocation, err = func() (*url.URL, error) { // A scope for defer
+		uploadReader := uploadreader.NewUploadReader(io.TeeReader(stream, io.MultiWriter(digester.Hash(), sizeCounter)))
+		// This error text should never be user-visible, we terminate only after makeRequestToResolvedURL
+		// returns, so there isn’t a way for the error text to be provided to any of our callers.
+		defer uploadReader.Terminate(errors.New("Reading data from an already terminated upload"))
+		res, err = d.c.makeRequestToResolvedURL(ctx, "PATCH", uploadLocation.String(), map[string][]string{"Content-Type": {"application/octet-stream"}}, uploadReader, inputInfo.Size, v2Auth, nil)
+		if err != nil {
+			logrus.Debugf("Error uploading layer chunked, response %#v", res)
+			return nil, err
+		}
+		defer res.Body.Close()
+		uploadLocation, err := res.Location()
+		if err != nil {
+			return nil, errors.Wrap(err, "Error determining upload URL")
+		}
+		return uploadLocation, nil
+	}()
 	if err != nil {
-		logrus.Debugf("Error uploading layer chunked, response %#v", res)
 		return types.BlobInfo{}, err
 	}
-	defer res.Body.Close()
 	computedDigest := digester.Digest()
-
-	uploadLocation, err = res.Location()
-	if err != nil {
-		return types.BlobInfo{}, errors.Wrap(err, "Error determining upload URL")
-	}
 
 	// FIXME: DELETE uploadLocation on failure (does not really work in docker/distribution servers, which incorrectly require the "delete" action in the token's scope)
 

--- a/internal/uploadreader/upload_reader.go
+++ b/internal/uploadreader/upload_reader.go
@@ -1,0 +1,61 @@
+package uploadreader
+
+import (
+	"io"
+	"sync"
+)
+
+// UploadReader is a pass-through reader for use in sending non-trivial data using the net/http
+// package (http.NewRequest, http.Post and the like).
+//
+// The net/http package uses a separate goroutine to upload data to a HTTP connection,
+// and it is possible for the server to return a response (typically an error) before consuming
+// the full body of the request. In that case http.Client.Do can return with an error while
+// the body is still being read — regardless of of the cancellation, if any, of http.Request.Context().
+//
+// As a result, any data used/updated by the io.Reader() provided as the request body may be
+// used/updated even after http.Client.Do returns, causing races.
+//
+// To fix this, UploadReader provides a synchronized Terminate() method, which can block for
+// a not-completely-negligible time (for a duration of the underlying Read()), but guarantees that
+// after Terminate() returns, the underlying reader is never used any more (unlike calling
+// the cancellation callback of context.WithCancel, which returns before any recipients may have
+// reacted to the cancellation).
+type UploadReader struct {
+	mutex sync.Mutex
+	// The following members can only be used with mutex held
+	reader           io.Reader
+	terminationError error // nil if not terminated yet
+}
+
+// NewUploadReader returns an UploadReader for an "underlying" reader.
+func NewUploadReader(underlying io.Reader) *UploadReader {
+	return &UploadReader{
+		reader:           underlying,
+		terminationError: nil,
+	}
+}
+
+// Read returns the error set by Terminate, if any, or calls the underlying reader.
+// It is safe to call this from a different goroutine than Terminate.
+func (ur *UploadReader) Read(p []byte) (int, error) {
+	ur.mutex.Lock()
+	defer ur.mutex.Unlock()
+
+	if ur.terminationError != nil {
+		return 0, ur.terminationError
+	}
+	return ur.reader.Read(p)
+}
+
+// Terminate waits for in-progress Read calls, if any, to finish, and ensures that after
+// this function returns, any Read calls will fail with the provided error, and the underlying
+// reader will never be used any more.
+//
+// It is safe to call this from a different goroutine than Read.
+func (ur *UploadReader) Terminate(err error) {
+	ur.mutex.Lock() // May block for some time if ur.reader.Read() is in progress
+	defer ur.mutex.Unlock()
+
+	ur.terminationError = err
+}

--- a/internal/uploadreader/upload_reader_test.go
+++ b/internal/uploadreader/upload_reader_test.go
@@ -1,0 +1,35 @@
+package uploadreader
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"io/ioutil"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestUploadReader(t *testing.T) {
+	// This is a smoke test in a single goroutine, without really testing the locking.
+
+	data := bytes.Repeat([]byte{0x01}, 65535)
+	// No termination
+	ur := NewUploadReader(bytes.NewReader(data))
+	read, err := ioutil.ReadAll(ur)
+	require.NoError(t, err)
+	assert.Equal(t, data, read)
+
+	// Terminated
+	ur = NewUploadReader(bytes.NewReader(data))
+	readLen := len(data) / 2
+	read, err = ioutil.ReadAll(io.LimitReader(ur, int64(readLen)))
+	require.NoError(t, err)
+	assert.Equal(t, data[:readLen], read)
+	terminationErr := errors.New("Terminated")
+	ur.Terminate(terminationErr)
+	read, err = ioutil.ReadAll(ur)
+	assert.Equal(t, terminationErr, err)
+	assert.Len(t, read, 0)
+}

